docs: distributed lock (#204)
GreptimeTeam/docs: docs/developer-guide/meta/dist-lock.md

# Distributed Lock

## Introduction

> `Meta` provides the functionality of distributed locks via gRPC.

Reviewer: ```suggestion
`Meta` provides the functionality of distributed locks via `gRPC`.
```

> Meta provides the ability to distribute lock externally via gRPC.

Reviewer: ```suggestion
`Meta` provides the functionality of distribute locks via gRPC.
```

Reviewer 2: since provides is used right underneath, can we change this sentence to sth like "by using gRPC, 'Meta' is able to distribute lock "?

## How to use

> A meta client implemented in rust is provided in [meta-client][1] crate.

Author: ```suggestion
A meta client implemented by rust is provided in [meta-client][1] crate.
```

> A meta client implemented by rust is provided in [meta-client][1] crate.

Reviewer: ```suggestion
The [meta-client][1] crate provides a Rust-based meta client implementation.
```

[1]: https://github.com/GreptimeTeam/greptimedb/tree/develop/src/meta-client

For example:

```rust
async fn do_some_work(meta_client: MetaClient) {

>     let name = "lock_name".as_bytes().to_vec();

Author: ```suggestion
    let name = b"lock_name".to_vec();
```

    let expire_secs = 60;

    let lock_req = LockRequest { name, expire_secs };

    let lock_result = meta_client.lock(lock_req).await.unwrap();
    let key = lock_result.key;

    info!("do some work, take 3 seconds");
    tokio::time::sleep(Duration::from_secs(3)).await;

    let unlock_req = UnlockRequest { key };

    meta_client.unlock(unlock_req).await.unwrap();
    info!("unlock success!");
}
```

More examples [here][2].

[2]: https://github.com/GreptimeTeam/greptimedb/blob/develop/src/meta-client/examples/lock.rs

> Because gRPC is cross-language, it is easy to implement clients for other programming languages.

Reviewer: @Fengys123 I can not understand the meaning of `cross-language` here. Can you give me some examples of those who also use this word?

> Because gRPC is cross-language, it is easy to implement clients for other programming languages.

Reviewer: ```suggestion
Because gRPC works across languages, it is easy to implement clients for other programming languages.
```

> For using other languages, since gRPC is cross-language, you can easily implement language-specific client.

Reviewer: ```suggestion
Because gRPC is cross-language, it is easy to implement clients for other programming languages.
```

> You can find the proto definition [here][3].

Reviewer: ```suggestion
You can find the protocol buffer definition [here][3].
```

[3]: https://github.com/GreptimeTeam/greptime-proto/blob/main/proto/greptime/v1/meta/lock.proto

> Pay attention to the following points

Reviewer 3: Pay attention to the following points
modify to
Pay attention to the following points:

> 1. If the expiration time of distribute lock is exceeded and currently holds the lock, the lock is automatically released.

Reviewer 2: ```suggestion
1. The distributed lock will be automatically released if the expiration time which holds it, is exceeded.
```

> 1. If the expiration time of distribute lock is exceeded and currently holds the lock, the lock is automatically released.

Reviewer: @apdong2022 it's not the expiration time holding the lock but the caller. A caller will request a lock with an expiration time, and the caller have to release the lock in time, otherwise the lock will be released by itself.

> 2. Distributed lock need to set a reasonable expiration time, default: 10 seconds.

Reviewer: ```suggestion
2. When using distributed locks, you must set a suitable expiration time, default: 10 seconds.
```

> 3. The expiration time of the distributed lock should be shorter than the timeout of the Grpc channel. Otherwise it may cause gRPC timeout error, causing lock failure.

Reviewer: ```suggestion
3. The expiration time of the distributed lock should be shorter than the timeout of the `gRPC` channel; otherwise it may cause `gRPC` timeout error, resulting in lock failure.
```

Reviewer 2: ```suggestion
3. Distributed lock's expiration time should be shorter than the 'gRPC' channel's timeout; otherwise, it may cause a 'gRPC' timeout error, resulting in lock failure.
```

## Detailed design

> The current implementation of distributed lock is based on [etcd][4]. And we define `Lock` trait, `EtcdLock` is one of the implementations.

Reviewer: ```suggestion
The current implementation of distributed lock is based on [etcd][4]. We have defined `Lock` trait, and `EtcdLock` is one of the available implementations.
```

[4]: https://etcd.io/docs/v3.5/dev-guide/api_concurrency_reference_v3/

> Since the state of the distributed lock is maintained by etcd, both meta leader and follower nodes can provide the Lock's Grpc service.

Reviewer: ```suggestion
Because the state of the distributed lock is maintained by `etcd`, both `meta` leader and follower nodes can provide the Lock's `gRPC` service.
```
